### PR TITLE
47 add timestamps

### PR DESCRIPTION
**Pull request by NickColley**

### CODE REVIEW

**Comment by NickColley** [comment]
In this PR I've reused the in-memory docs transform so we don't have to to a PouchDB#get() however this is a point of discussion since the code maybe simpler if we did use it since now we're merging three objects:
- docs
- object
- response

**Comment by NickColley** [comment]
Looks like saucelabs timed out

**Comment by NickColley** [comment]
@gr2m note if we merge https://github.com/hoodiehq/pouchdb-hoodie-api/pull/71 first this'll need to be updated.

File: utils/to-doc.js [hunk 1]
```diff
@@ -1,12 +1,16 @@
...
 var extend = require('pouchdb-extend')
+var now = require('./now')
 
 module.exports = function objectToDoc (object) {
   var doc = extend({}, object, {
     _id: object.id
   })
 
+  doc.updatedAt = now()
+  doc.createdAt = doc.createdAt || doc.updatedAt
```
Comment (gr2m) [inline utils/to-doc.js]: I'm not convinced that this is the place to add the timestamps. I think we should add a dedicated method like "addTimestamps" and call it explicitly. What do you think @nickcolley?

Comment (NickColley) [inline utils/to-doc.js]: I figured this transform was taking an object and creating a 'hoodie-style doc' so timestamps are typical of that, I'd be okay with an explicit method if it makes it more obvious.

Comment (gr2m) [inline utils/to-doc.js]: Ah okay, `toDoc` is actually the other way, taking a CouchDB `doc` and turning it into a hoodie-esque `object`, so [utils/to-object.js](utils/to-object.js) might be the better place to put this

Comment (NickColley) [inline utils/to-doc.js]: Trouble is that is used after we've got a response but we need to store the timestamps, so maybe let's go with the dedicated method.

Comment (gr2m) [inline utils/to-doc.js]: yeah, got it. I agree, let's do a dedicated method

**Comment by gr2m** [comment]
> note if we merge #71 first this'll need to be updated

it's merged now

**Comment by NickColley** [comment]
Suspect lolex doesnt work in the browser maybe..

**Comment by gr2m** [comment]
I tried it in chrome, works fine:

``` js
lolex.install()
new Date
// Wed Dec 31 1969 19:00:00 GMT-0500 (EST)
```

I've restarted the build, let's see ... :pray:

**Comment by NickColley** [comment]
Lolex.install() stubs timers and dates, I've pulled in Sinon and told it to only stub Date. :+1:

File: tests/specs/add.js [hunk 2]
```diff
@@ -132,3 +133,57 @@ test('adds multiple objects to db', function (t) {
     t.is(res.total_rows, 3, 'puts docs')
   })
 })
+
+test('store.add(object) makes createdAt and updatedAt timestamps', function (t) {
+  t.plan(4)
+
+  var clock = sinon.useFakeTimers('Date')
```
Comment (gr2m) [inline tests/specs/add.js]: Do you know how this works internally? Maybe we don't need to pull in the entire `sinon` package, if faking date is the only thing we need?

File: helpers/update-many.js [hunk 3]
```diff
@@ -4,13 +4,17 @@ var extend = require('pouchdb-extend')
...
 var changeObject = require('../utils/change-object')
 var toDoc = require('../utils/to-doc')
+var addTimestamps = require('../utils/add-timestamps')
+var getTimestamps = require('../utils/get-timestamps')
 var toId = require('../utils/to-id')
+
 var findMany = require('./find-many')
 
 module.exports = function updateMany (array, change) {
   var self = this
```
Comment (varjmes) [inline helpers/update-many.js:13]: Would:

``` javascript
   var self = this,
       errors = this.constructor.Errors,
       objects,
       docs,
       ids = array.map(toId);
```

Be nicer here? :)

File: helpers/update-one.js [hunk 4]
```diff
@@ -11,6 +14,7 @@ module.exports = function updateOne (idOrObject, change) {
   var Promise = this.constructor.utils.Promise
```
Comment (varjmes) [inline helpers/update-one.js:13]: What about writing:

``` javascript
   var Promise = this.constructor.utils.Promise,
       errors = this.constructor.Errors,
       object,
       doc;
```

File: remove-all.js [hunk 5]
```diff
@@ -15,6 +17,7 @@ module.exports = removeAll
  */
 function removeAll (filter) {
   var objects
+  var docs
```
Comment (varjmes) [inline remove-all.js]:
``` javascript
var objects,
    docs;
```

File: update-all.js [hunk 6]
```diff
@@ -20,6 +22,7 @@ function updateAll (changedProperties) {
 
   var type = typeof changedProperties
```
Comment (varjmes) [inline update-all.js:22]:
``` javascript
   var type = typeof changedProperties,
       objects,
       docs;
```

File: helpers/update-many.js (hunk 3 above)
Comment (gr2m) [inline helpers/update-many.js:13]: we do one `var` per variable in all other repos, I'd keep that consistent

**Comment by gr2m** [comment]
I've a conceptual question, maybe we should put this into it's own issue: when I remove an object, without updating it, should we set `updatedAt` to now, or should we only set `removedAt`?

File: helpers/update-many.js (hunk 3 above)
Comment (NickColley) [inline helpers/update-many.js:13]: I prefer it also, no comma juggling.

**Comment by gr2m** [comment]
I'm a bit confused by the two methods, `add-timestamps` and `get-timestamps`. Can you explain why we need both? 

For example, here's a simplified `add-many.js`:

``` js
var toDoc = require('../utils/to-doc')
var addTimestamps = require('../utils/add-timestamps')
// var getTimestamps = require('../utils/get-timestamps')

module.exports = function addMany (objects) {
  objects.forEach(addTimestamps)
  var docs = objects.map(toDoc)

  return this.bulkDocs(docs)

  .then(function (responses) {
    return responses.map(function (response, i) {
      if (response instanceof Error) {
        return response
      }

      objects[i].id = response.id
      objects[i]._rev = response.rev
      // return getTimestamps(objects[i], docs[i])
      return objects[i]
    })
  })
}
```

File: helpers/add-many.js [hunk 7]
```diff
@@ -1,11 +1,11 @@
 'use strict'
 
 var toDoc = require('../utils/to-doc')
+var addTimestamps = require('../utils/add-timestamps')
 
 module.exports = function addMany (objects) {
-  var docs = objects.map(toDoc)
```
Comment (NickColley) [inline helpers/add-many.js:6]: Think this original docs set me on that path, stopped me from seeing the simpler solution.

**Comment by gr2m** [comment]
merged via 3b944ed

**Pull request closed without merging**